### PR TITLE
remove annoying steps without input but informations to "continue"

### PR DESCRIPTION
**Pull request by adietish**
# DO NOT MERGE, PLEASE PROVIDE FEEDBACK

This is an imperfect suggestion to get rid of the annoying steps when providing user inputs where one gets an info and simply has to hit "[continue]". 
These are annoying and misleading, it's too easy to try to provide your username and then notice that no, you just have to hit "continue". Providing the username is in the next step.

steps:
1. EXEC: pick "Download Runtime" in context menu of the "Servers" view
2. ASSERT: list of available runtimes is presented in quick pick list
3. EXEC: choose JBoss EAP 7.0 (or any other EAP)

Result:
image
You get a quick pick where there's just an info being displayed which one has to confirm with "continue"

### CODE REVIEW

**Comment by adietish** [comment]
@dgolovin thoughts?

**Comment by robstryker** [comment]
The server will send a list of workflow items. Some of these items will be simply what we would consider labels in UI. A text label with no ability for the user to change anything. 

Your patch here is to basically refuse to show any labels that don't require a response.

I simply cannot understand in what world you think this is a valid solution. If the server type demands that a user see a label or message, even one that doesn't require a response, then the client must show it to them.

**Comment by adietish** [comment]
@dgolovin looks like we have to display all steps. Isn't there a better way to display those information-only steps? Currently quickPick is used, the server provided label is in the input-field. Isn't there a better way to display those infos without an input-field?

File: src/extensionApi.ts [hunk 1]
```diff
@@ -233,37 +239,38 @@ export class CommandHandler {
...
     }
 
-    private async promptUser(item: Protocol.WorkflowResponseItem, workflowMap: {}): Promise<boolean> {
-        const prompt = item.label + (item.content ? `\n${item.content}` : '');
+    private async promptUser(prompt: string, item: Protocol.WorkflowResponseItem, workflowMap: {}): Promise<boolean> {
         let userInput: any = null;
-        if (item.responseType === 'none') {
-            userInput = await vscode.window.showQuickPick(['continue...'], { placeHolder: prompt, ignoreFocusOut: true });
+        if (item.responseType === ResponseType.BOOL) {
+            const oneProp = await vscode.window.showQuickPick(['true', 'false'], { placeHolder: prompt, ignoreFocusOut: true });
```
Comment (jrichter1) [inline src/extensionApi.ts]: 'Yes'/'No' would be more appreciated I think.

File: src/extensionApi.ts [hunk 2]
```diff
@@ -233,37 +239,38 @@ export class CommandHandler {
...
+            userInput = (oneProp === 'true');
         } else {
-            if (item.responseType === 'bool') {
-                const oneProp = await vscode.window.showQuickPick(['true', 'false'], { placeHolder: prompt, ignoreFocusOut: true });
-                userInput = (oneProp === 'true');
+            const oneProp = await vscode.window.showInputBox({ prompt: prompt, ignoreFocusOut: true });
```
Comment (jrichter1) [inline src/extensionApi.ts]: A case for password masking is sorely needed here.

File: src/extensionApi.ts [hunk 3]
```diff
@@ -233,37 +239,38 @@ export class CommandHandler {
...
+            if (item.responseType === ResponseType.INT) {
+                userInput = +oneProp;
```
Comment (jrichter1) [inline src/extensionApi.ts]: Cancelling an int prompt gets you ```userInput = NaN``` which makes the method return false.

**Comment by jrichter1** [comment]
> @dgolovin We currently display simple text messages with quickPicks and a possible choice of "continue" in the selectable options. Isn't there a better way to display those infos without an input-field?

You could use an info/warning message - they actually let you add custom buttons and can be made modal if needed.

**Comment by adietish** [comment]
> > @dgolovin We currently display simple text messages with quickPicks and a possible choice of "continue" in the selectable options. Isn't there a better way to display those infos without an input-field?
> 
> You could use an info/warning message - they actually let you add custom buttons and can be made modal if needed.
> https://code.visualstudio.com/api/references/vscode-api#window.showInformationMessage



> > @dgolovin We currently display simple text messages with quickPicks and a possible choice of "continue" in the selectable options. Isn't there a better way to display those infos without an input-field?
> 
> You could use an info/warning message - they actually let you add custom buttons and can be made modal if needed.
> https://code.visualstudio.com/api/references/vscode-api#window.showInformationMessage

@jrichter1 Aren't information messages shown in the bottom right corner which then would lead to weird behaviour where a 1st message is show bottom right, the input prompt for the username then on the upper edge etc.?

**Comment by jrichter1** [comment]
> @jrichter1 Aren't information messages shown in the bottom right corner which then would lead to weird behaviour where a 1st message is show bottom right, the input prompt for the username then on the upper edge etc.?

Correct. I was going to say there was no other way to get input, but it seems something new appeared in the API recently. I haven't tried it out, but it might be worth checking.
https://code.visualstudio.com/api/references/vscode-api#window.createInputBox

**Comment by lstocchi** [comment]
What i find annoying of these prompts is that when I type something the quickpick start filtering the options available even though it only has one single option "continue". In this case, if a user doesn't pay attention and type a letter which is not contained in the word "continue", the option disappear, and user can be puzzled.

Even though the API added some new features (as mentioned by @jrichter1 ) i haven't found a solution which completely satisfy me. 

The only thing I would change from the current implemention is using the property "alwaysShow" which keeps the option always visible (no matter what user types) and the event "onDidChangeValue" that we can use to display a warning message to tell the user what he has to do. 

https://code.visualstudio.com/api/references/vscode-api#QuickPick%3CT%3E

**Comment by odockal** [comment]
@adietish Is this still actual? It would be nice to have this done. Needs rebase in any case.

**Comment by adietish** [comment]
@odockal the core in this PR is nothing but a POC that's far from being useable. Maybe @lstocchi can look into this?

**Comment by lstocchi** [comment]
PR moved to https://github.com/redhat-developer/vscode-rsp-ui/pull/76

**Pull request closed without merging**